### PR TITLE
bugfix: add support for `global_ordinal`, `local_ordinal`, `world_size` in xla

### LINKED ISSUE #20852: Support PyTorch/XLA 2.7
### Description & Motivation

[PyTorch / XLA 2.7](https://github.com/pytorch/xla/releases/tag/v2.7.0) deprecates several APIs that Lightning still relies on. Could Lightning be updated for full XLA 2.7 compatibility?

### Pitch

_No response_

### Alternatives

_No response_

### Additional context

_No response_

cc @lantiga @borda

### PR DESCRIPTION
**Pull request by AlexandrByzov**
## What does this PR do?

PyTorch XLA has deprecated several methods in PyTorch 2.7:
- Remove torch_xla.core.xla_model.xrt_world_size, replace with torch_xla.runtime.world_size
- Remove torch_xla.core.xla_model.get_ordinal, replace with torch_xla.runtime.global_ordinal

See [here](https://github.com/pytorch/xla/releases)

Fixes #20852 

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs): No
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs): Tests were already written
- [x] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

----
📚 Documentation preview 📚: https://pytorch-lightning--20872.org.readthedocs.build/en/20872/

### CODE REVIEW

**Review by Borda — COMMENTED** [review]
thank you for this finding!
could you pls add test to cover this new behaviour?
can this XLA be also run without TPU?

**Comment by AlexandrByzov** [comment]
> thank you for this finding! could you pls add test to cover this new behaviour? can this XLA be also run without TPU?

Thank you for taking the time to review. I implemented a test with mocks, because as far as I understand `torch_xla` is not used in tests without TPU. Generally, it is very much possible to use XLA with CPU or even GPU, but this will require a large change in tests and probably a separate issue/s

File: tests/tests_fabric/plugins/environments/test_xla.py [hunk 1]
```diff
@@ -97,3 +97,31 @@ def test_detect(monkeypatch):
 
     monkeypatch.setattr(lightning.fabric.accelerators.xla.XLAAccelerator, "is_available", lambda: True)
     assert XLAEnvironment.detect()
+
+
+@mock.patch.dict(os.environ, {}, clear=True)
+def test_attributes_from_xla_greater_21_used(xla_available, monkeypatch):
+    """Test XLA environment attributes when using XLA runtime >= 2.1."""
+    monkeypatch.setattr(lightning.fabric.accelerators.xla, "_XLA_GREATER_EQUAL_2_1", True)
+    monkeypatch.setattr(lightning.fabric.plugins.environments.xla, "_XLA_GREATER_EQUAL_2_1", True)
```
Comment (bhimrazy) [inline tests/tests_fabric/plugins/environments/test_xla.py]:
```suggestion


@mock.patch.dict(os.environ, {}, clear=True)
@mock.patch("lightning.fabric.accelerators.xla._XLA_GREATER_EQUAL_2_1", True)
@mock.patch("(lightning.fabric.plugins.environments.xla._XLA_GREATER_EQUAL_2_1", True)
def test_attributes_from_xla_greater_21_used(xla_available, monkeypatch):
    """Test XLA environment attributes when using XLA runtime >= 2.1."""
```

File: src/lightning/fabric/plugins/environments/xla.py [hunk 2]
```diff
@@ -66,6 +66,11 @@ def world_size(self) -> int:
         The output is cached for performance.
 
         """
+        if _XLA_GREATER_EQUAL_2_1:
+            from torch_xla import runtime as xr
+
+            return xr.world_size()
+
```
Comment (bhimrazy) [inline src/lightning/fabric/plugins/environments/xla.py:73]: Thanks for the update! Just to clarify — shouldn't this be `_XLA_GREATER_EQUAL_2_7` instead? Pls let me know if I’m missing something here.

File: tests/tests_fabric/plugins/environments/test_xla.py [hunk 3]
```diff
@@ -97,3 +97,31 @@ def test_detect(monkeypatch):
 
     monkeypatch.setattr(lightning.fabric.accelerators.xla.XLAAccelerator, "is_available", lambda: True)
     assert XLAEnvironment.detect()
+
+
+@mock.patch.dict(os.environ, {}, clear=True)
+@mock.patch("lightning.fabric.accelerators.xla._XLA_GREATER_EQUAL_2_1", True)
+@mock.patch("(lightning.fabric.plugins.environments.xla._XLA_GREATER_EQUAL_2_1", True)
```
Comment (bhimrazy) [inline tests/tests_fabric/plugins/environments/test_xla.py]:
```suggestion
@mock.patch("lightning.fabric.plugins.environments.xla._XLA_GREATER_EQUAL_2_1", True)
```
my bad

File: src/lightning/fabric/plugins/environments/xla.py (hunk 2 above)
Comment (AlexandrByzov) [reply src/lightning/fabric/plugins/environments/xla.py:73]: Thank you very much for the review and suggestions!

Yes, the old methods were deprecated in 2.7, but the new methods were introduced already in 2.1. You could see it here:  https://github.com/pytorch/xla/releases/tag/v2.1.0, MR#5011
1. `world_size`: https://github.com/pytorch/xla/pull/5011/files#diff-b5151ded18fbfdf1debacf597819bd88f7954b7f67f951fa8e131340805e5df5R129
2. `global_ordinal`: https://github.com/pytorch/xla/pull/5011/files#diff-b5151ded18fbfdf1debacf597819bd88f7954b7f67f951fa8e131340805e5df5R152
3. `local_ordinal`: https://github.com/pytorch/xla/pull/5011/files#diff-b5151ded18fbfdf1debacf597819bd88f7954b7f67f951fa8e131340805e5df5R162

I did not want to introduce more flags when the current ones should do the job, hence this usage of 2.1.

I also just now ran this on [Google Collab](https://colab.research.google.com/drive/13d6f3SCJBwOJaVn8UP-Nm9xWqCZLRGgk?usp=sharing) just to check if there were any differences in version 2.1, but it seems that the outputs are identical:

```python
import torch_xla
from torch_xla import runtime as xr
import torch_xla.core.xla_model as xm

print("TORCH XLA VERSION: ", torch_xla.__version__)

assert xr.world_size() == xm.xrt_world_size()
assert xr.local_ordinal() == xm.get_local_ordinal()
assert xr.global_ordinal() == xm.get_ordinal()
```

Comment (bhimrazy) [reply src/lightning/fabric/plugins/environments/xla.py:73]: Oh, got it! Thanks a lot for the clarification, @AlexandrByzov. 
I mistakenly assumed those methods got introduced in 2.7 — my bad.

**Review by bhimrazy — APPROVED** [review]

File: src/lightning/fabric/plugins/environments/xla.py (hunk 2 above)
Comment (AlexandrByzov) [reply src/lightning/fabric/plugins/environments/xla.py:73]: Thank you for adding this question, it would provide a context to this decision for others too!

**Comment by AlexandrByzov** [comment]
Is there anything else that can be done for this MR @Borda ? I will be glad to work on it further

**Review by deependujha — APPROVED** [review]

File: tests/tests_fabric/plugins/environments/test_xla.py [hunk 4]
```diff
@@ -97,3 +97,31 @@ def test_detect(monkeypatch):
...
+
+        assert env.world_size() == 4
+        assert env.global_rank() == 2
+        assert env.local_rank() == 1
+
+        env.set_world_size(100)
+        assert env.world_size() == 4
+
+        env.set_global_rank(100)
+        assert env.global_rank() == 2
```
Comment (Borda) [inline tests/tests_fabric/plugins/environments/test_xla.py]: can we pls split it into smaller tests

Comment (bhimrazy) [reply tests/tests_fabric/plugins/environments/test_xla.py]: cc: @AlexandrByzov

Comment (bhimrazy) [reply tests/tests_fabric/plugins/environments/test_xla.py]: Addressing the feedback.

File: src/lightning/fabric/CHANGELOG.md [hunk 5]
```diff
@@ -16,11 +16,17 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
...
 
+
 ### Changed
 
 - Raise ValueError when seed is `out-of-bounds` or `cannot be cast to int` ([#21029](https://github.com/Lightning-AI/pytorch-lightning/pull/21029))
 
 
+### Fixed
+
+- Fix XLA strategy to add support for  for global_ordinal, local_ordinal, world_size which came instead of deprecated methods ([#20852](https://github.com/Lightning-AI/pytorch-lightning/issues/20852))
```
Comment (Borda) [inline src/lightning/fabric/CHANGELOG.md]:
```suggestion
- Fix XLA strategy to add support for `global_ordinal`, `local_ordinal`, `world_size` which came instead of deprecated methods ([#20852](https://github.com/Lightning-AI/pytorch-lightning/issues/20852))
```

**Review by Borda — APPROVED** [review]
